public/v2/apps/directus.json

```diff
@@ -24,48 +24,41 @@
-                    "ADMIN_EMAIL": "$$cap_directus_admin_username",
-                    "ADMIN_PASSWORD": "$$cap_directus_admin_passwd"
-                }
-            },
-            "$$cap_appname-app": {
-                "depends_on": {
-                    "$$cap_appname-db": {
-                        "condition": "service_healthy"
-                    },
-                    "$$cap_appname-api": {}
-                },
-                "image": "directus/app:7.0.18",
-                "containerHttpPort": "80",
-                "restart": "always",
-                "environment": {
-                    "API_ENDPOINT": "API; $$cap_api_endpoint/_/"
+                    "DIRECTUS_DATABASE_HOST": "srv-captain--$$cap_appname-db",
+                    "DIRECTUS_DATABASE_NAME": "$$cap_mysql_database_name",
+                    "DIRECTUS_DATABASE_USERNAME": "$$cap_mysql_user",
+                    "DIRECTUS_DATABASE_PASSWORD": "$$cap_mysql_passwd",
+                    "DIRECTUS_AUTH_SECRETKEY": "$$cap_directus_auth_secret_key",
+                    "DIRECTUS_AUTH_PUBLICKEY": "$$cap_directus_auth_public_key"
                 }
             }
         }
     },
     "instructions": {
-        "end": "Directus is deployed and available as $$cap_appname-app. \n\n It may take a few minutes for Directus to initialize.\n",
-        "start": "Directus is a free and open souce content management system and API wrapper for SQL databases. The API instance provides a realtime REST API for interacting with your database, while the App instance provides a web based CMS for managaging the database. \n For more information and available options, see https://github.com/directus/docker."
+        "end": "THE INSTALLATION PROCESS IS NOT FINISHED !!!\n\nAs said in the documentation (https://docs.directus.io/installation/docker.html#step-4-initialize-the-database-and-an-admin-user), currently (30/06/2020), Directus does not support automatic database setup so you need to do it yourself the first time you launch directus.\n\n But since Caprover does not use standard docker-compose file, you need to run a slightly uglier command than the one shown in the documentation.\n\n Run the following command on your server (using SSH for example, you might need to be root):\n\ndocker run -e DIRECTUS_DATABASE_HOST=srv-captain--$$cap_appname-db -e DIRECTUS_DATABASE_NAME=$$cap_mysql_database_name -e DIRECTUS_DATABASE_USERNAME=$$cap_mysql_user -e DIRECTUS_DATABASE_PASSWORD=$$cap_mysql_passwd -e DIRECTUS_AUTH_SECRETKEY=$$cap_directus_auth_secret_key -e DIRECTUS_AUTH_PUBLICKEY=$$cap_directus_auth_public_key --network $(docker network ls | grep \"captain-overlay-network\" | awk '{print $1}') --rm directus/directus:$$cap_directus_tag install --email $$cap_directus_install_email --password $$cap_directus_install_password\n\nOnce this is done, Directus is deployed and available as $$cap_appname-directus.\n\n It may take a few minutes for Directus to initialize.\n",
```

Comment:
This works, but it's a bit ugly. How about making our own installer:
- Add another service: `$$cap_appname-directus`
- Set dockerfileLines of that service to  to 
```
FROM directus/directus:$$cap_directus_tag
ENTRYPOINT ["/bin/sh"]
CMD directus-entrypoint install --email $$cap_directus_install_email --password $$cap_directus_install_password && sleep 9999d
```

Then in instructions.end we can direct user to simply remove this service after a few minutes.


I haven't tested this method, but I think something like this should work. What do you thing?

Reply:
What you propose is a good idea. I have tried to implement it for the past 2 to 3 hours now and I always end up with the same problem.

I have added the following `installer` service :
```json
"$$cap_appname-installer": {
    "depends_on": {
        "$$cap_appname-db": {
            "condition": "service_healthy"
        },
        "$$cap_appname-directus": {}
    },
    "notExposeAsWebApp": "true",
    "dockerfileLines": [
        "FROM directus/directus:$$cap_directus_tag",
        "ENTRYPOINT [\"/bin/bash\"]",
        "CMD [\"directus-entrypoint\",\"install\", \"--email\", \"$$cap_directus_install_email\", \"--password\", \"$$cap_directus_install_password\", \"&&\", \"sleep\", \"9999d\"]"
    ],
    "environment": {
        "DIRECTUS_DATABASE_HOST": "srv-captain--$$cap_appname-db",
        "DIRECTUS_DATABASE_NAME": "$$cap_mysql_database_name",
        "DIRECTUS_DATABASE_USERNAME": "$$cap_mysql_user",
        "DIRECTUS_DATABASE_PASSWORD": "$$cap_mysql_passwd",
        "DIRECTUS_AUTH_SECRETKEY": "$$cap_directus_auth_secret_key",
        "DIRECTUS_AUTH_PUBLICKEY": "$$cap_directus_auth_public_key"
    },
    "restart": "no"
}
```

It is able to create `directus` tables successfully but it never takes the `email` and the `password` I give it, even when they are hard codded. If I run `select * from directus_users;` in my database, I get this : 
```bash
mysql> select * from directus_users;
+----+--------+------+------------+-----------+-------+--------------------------------------------------------------+-------+----------+--------+----------------+--------+---------+-------+---------------------+----------------+-----------+-------------+-------+------------+----------------------+
| id | status | role | first_name | last_name | email | password                                                     | token | timezone | locale | locale_options | avatar | company | title | email_notifications | last_access_on | last_page | external_id | theme | 2fa_secret | password_reset_token |
+----+--------+------+------------+-----------+-------+--------------------------------------------------------------+-------+----------+--------+----------------+--------+---------+-------+---------------------+----------------+-----------+-------------+-------+------------+----------------------+
|  1 | active |    1 | Admin      | User      |       | $2y$10$pofjlDvWJ7rr1lVVDHqaluTcOrm1NIZlCGVKWRVpzGsfLyHCePWV6 | NULL  | UTC      | en-US  | NULL           |   NULL | NULL    | NULL  |                   1 | NULL           | NULL      | NULL        | auto  | NULL       | NULL                 |
+----+--------+------+------------+-----------+-------+--------------------------------------------------------------+-------+----------+--------+----------------+--------+---------+-------+---------------------+----------------+-----------+-------------+-------+------------+----------------------+
1 row in set (0.00 sec)
```

I have even tried to run it without `email` and `password` parameters or to use environment variables but it does not change a thing : 
```json
"dockerfileLines": [
    "FROM directus/directus:$$cap_directus_tag",
    "ENTRYPOINT [\"/bin/bash\"]",
    "CMD [\"directus-entrypoint\", \"install\", \"--email\", \"$DIRECTUS_INSTALL_EMAIL\", \"--password\", \"$DIRECTUS_INSTALL_PASSWORD\", \"&&\", \"sleep\", \"9999d\"]"
],
"environment": {
    "DIRECTUS_DATABASE_HOST": "srv-captain--$$cap_appname-db",
    "DIRECTUS_DATABASE_NAME": "$$cap_mysql_database_name",
    "DIRECTUS_DATABASE_USERNAME": "$$cap_mysql_user",
    "DIRECTUS_DATABASE_PASSWORD": "$$cap_mysql_passwd",
    "DIRECTUS_AUTH_SECRETKEY": "$$cap_directus_auth_secret_key",
    "DIRECTUS_AUTH_PUBLICKEY": "$$cap_directus_auth_public_key",
    "DIRECTUS_INSTALL_EMAIL": "$$cap_directus_install_email",
    "DIRECTUS_INSTALL_PASSWORD": "$$cap_directus_install_password"
}
```
```json
"dockerfileLines": [
    "FROM directus/directus:$$cap_directus_tag",
    "ENTRYPOINT [\"/bin/bash\"]",
    "CMD [\"directus-entrypoint\", \"install\", \"&&\", \"sleep\", \"9999d\"]"
],
"environment": {
    "DIRECTUS_DATABASE_HOST": "srv-captain--$$cap_appname-db",
    "DIRECTUS_DATABASE_NAME": "$$cap_mysql_database_name",
    "DIRECTUS_DATABASE_USERNAME": "$$cap_mysql_user",
    "DIRECTUS_DATABASE_PASSWORD": "$$cap_mysql_passwd",
    "DIRECTUS_AUTH_SECRETKEY": "$$cap_directus_auth_secret_key",
    "DIRECTUS_AUTH_PUBLICKEY": "$$cap_directus_auth_public_key",
    "DIRECTUS_INSTALL_EMAIL": "$$cap_directus_install_email",
    "DIRECTUS_INSTALL_PASSWORD": "$$cap_directus_install_password"
}
```

Also ;
- the password hash seems to change all the time, so either directus uses a salt (in this case it is perfectly normal) or the password is randomly generated meaning that the password I provide is never used 
- the `sleep` command in the `dockerfileLines` does not work and the `restart: no` neither. Indeed, if I check the logs in `caprover` I can see that the command `directus-entrypoint ...` runs over and over.

I am going to stop working on that since I do not really have the time. If anyone want to try to fix this, you are more than welcome. In the meantime, my first proposed solution works (even if, I totally agree, it is a bit ugly) and this is the one I am going to use for now.

Reply:
I agree. We'll get to fixing it later.